### PR TITLE
Reassign MME cases to another user when a Scout user is removed

### LINKED ISSUE #2979: Reassigning MatchMaker Exchange cases to another user when a user leaves
If a user is leaving and is the contact for some Patients, reassign these patients to another user. This is to address a comment from a reviewer of out PatientMatcher publication.

### PR DESCRIPTION
**Pull request by northwestwitch**
Fix #2979

**How to test**:
1. In a terminal, start a MatchMaker Exchange server with demo data:
```
git clone https://github.com/Clinical-Genomics/patientMatcher.git
cd patientMatcher
docker-compose up
```
2. In a local instance of Scout, demo config file (config.py), uncomment these lines: 
```
MME_ACCEPTS = "application/vnd.ga4gh.matchmaker.v1.0+json"
MME_URL = "http://localhost:8000"
MME_TOKEN = "DEMO"
```
3. Modify demo user and grant him permission to share cases to MatchMaker:
```
scout --demo update user -u clark.kent@mail.com -r mme_submitter
```
4. Launch the web app of Scout, pin a variant and share case to the MatchMaker.
image

5. Check in MongoDB compass, where a submission object dictionary should have been created for the demo case. Note that the submitter **Clark Kent is present twice in this object: both in patient.contact as an object and its mail under subm_user**

image

6. Stop the Scout server and try to delete the demo user with this command:
```
scout --demo delete user -m clark.kent@mail.com
```
7. It should complain and ask you if you want to reassign the MME cases:

image

8. If you type N, then the command should exit.
9. If you type Y, the it will ask you the email for the new MME contact, a user already present in Scout.
10. Create a new user and repeat, try to assign the MME case to this user. It should fail if the new user doesn't have the "mme_submitter" role
11. After updating the user and granting them the `mme_submitter` (as in 3), the command should be successful and:
- The case should be reassigned to the new user (check the case document in mongo compass)
- Demo user Clark Kent should have been removed.

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by
- [x] tests executed by

### CODE REVIEW

**Comment by northwestwitch** [comment]
1. If the case is not assigned to another user, then demo user Clark Kent can't be removed:
image

2. Assign the case to a user with email not present in Scout:
image

3. Create a new user that is not a MME submitter and use it for reassigning the case:
image

image

4. Update new user with mme_submitter role:
image

5. Try again to remove demo user Clark Kent:
image

User is removed from database:
image

Case submission is updated:
image

A new event is created when the MME contact is modified:
image

**Comment by northwestwitch** [comment]
This is ready for review now!

**Comment by northwestwitch** [comment]
I'm working on the PatientMatcher paper revision and I'm including this feature in the text. This means we should try to merge this PR as soon as possible..

**Comment by mikaell** [comment]
I think it should be more explicit that the user will be deleted only if the matchmaker is reassigned.

```
(py3.8) Mikael@mikaellmbp:scout% scout delete user -m ml@mail.com
2021-12-16 09:55:05 mikaellmbp scout.commands.base[75670] INFO Running scout version 4.45
2021-12-16 09:55:05 mikaellmbp scout.commands.base[75670] DEBUG Debug logging enabled.
2021-12-16 09:55:05 mikaellmbp scout.adapter.client[75670] INFO Connected to MongoDB 4.0.3
2021-12-16 09:55:05 mikaellmbp scout.adapter.mongo.base[75670] DEBUG Database name: scout
2021-12-16 09:55:05 mikaellmbp scout.commands.delete.delete_command[75670] INFO Running scout delete user
2021-12-16 09:55:05 mikaellmbp scout.adapter.mongo.user[75670] INFO Fetching user ml@mail.com
2021-12-16 09:55:05 mikaellmbp scout.adapter.mongo.case[75670] INFO Fetching case internal_id
MatchMaker Exchange submissions (n. cases=1) are associated to this user. Reassign patients to another user? [y/N]: N
Aborted!
```

File: scout/commands/delete/delete_command.py [hunk 1]
```diff
@@ -227,6 +227,23 @@ def user(mail):
         LOG.warning("User {0} could not be found in database".format(mail))
         return
 
+    # Check if user has associated MatchMaker Exchange submissions
+    mme_submitted_cases = adapter.user_mme_submissions(user_obj)
+    if mme_submitted_cases:
+        click.confirm(
+            f"MatchMaker Exchange submissions (n. cases={len(mme_submitted_cases)}) are associated to this user. Reassign patients to another user?",
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/commands/delete/delete_command.py:234]: Something like this @mikaell ?

[the remark this answers is not in the document]
Comment (mikaell) [reply scout/commands/delete/delete_command.py:234]: Perfect!

File: scout/adapter/mongo/matchmaker.py [hunk 2]
```diff
@@ -8,6 +8,64 @@
...
+        """
+        # All events associated to the provided user
+        user_events = self.user_events(user_obj)
+        # Collect unique case _id with MME submissions from provided user
+        submitted_cases = set()
+
+        for event in user_events:
+            if event.get("verb") != "mme_add":
+                continue
+            # Check that the cases patient's submission are stll actual
```
Comment (mikaell) [inline scout/adapter/mongo/matchmaker.py:27]: Typo: still

**Review by mikaell — CHANGES_REQUESTED** [review]
Looks good! 

But I found a typo. I should be fixed as it has been spotted 🙂

**Review by mikaell — APPROVED** [review]

**Comment by northwestwitch** [comment]
Thank you, merging now!

**Pull request merged**